File: dataprep-transformation/src/main/resources/TDE_API/share/doc/dataextract-8300.15.0308.1149/samples/MakeOrder.cpp
// -----------------------------------------------------------------------
// The information in this file is the property of Tableau Software and
// is confidential.
//
// Copyright (C) 2012  Tableau Software.
// Patents Pending.
// -----------------------------------------------------------------------
// externalapi/samples/MakeOrder.cpp
// -----------------------------------------------------------------------
#include "DataExtract_cpp.h"

#include <iomanip>
#include <iostream>
#include <memory>
#include <stdlib.h>

// Define the table's schema
void MakeTableDefinition( Tableau::TableDefinition& tableDef )
{
    using namespace Tableau;
    tableDef.SetDefaultCollation( Collation_en_GB );
    tableDef.AddColumn( L"Purchased",       Type_DateTime );
    tableDef.AddColumn( L"Product",         Type_CharString );
    tableDef.AddColumn( L"uProduct",        Type_UnicodeString );
    tableDef.AddColumn( L"Price",           Type_Double );
    tableDef.AddColumn( L"Quantity",        Type_Integer );
    tableDef.AddColumn( L"Taxed",           Type_Boolean );
    tableDef.AddColumn( L"Expiration Date", Type_Date );

    // Override default collation
    tableDef.AddColumnWithCollation( L"Produkt", Type_CharString, Collation_de );
}

// Print a Table's schema to stderr.
void PrintTableDefinition( Tableau::TableDefinition& tableDef )
{
    const int numColumns = tableDef.GetColumnCount();
    for ( int i = 0; i < numColumns; ++i ) {
        Tableau::Type type = tableDef.GetColumnType(i);
        std::wstring name = tableDef.GetColumnName(i);
        std::wcerr << L"Column " << i << L": " << name << L" ("
                   << std::hex << std::showbase << std::internal << std::setfill(L'0') << std::setw(6)
                   << type << std::dec << L")\n";
    }
}

// Insert a few rows of data.
void InsertData( std::shared_ptr<Tableau::Table> table )
{
    std::shared_ptr<Tableau::TableDefinition> tableDef = table->GetTableDefinition();
    Tableau::Row row( *tableDef );

    row.SetDateTime(  0, 2012, 7, 3, 11, 40, 12, 4550 ); // Purchased
    row.SetCharString(1, "Beans" );                      // Product
    row.SetString(    2, L"uniBeans" );                  // unicode Product
    row.SetDouble(    3, 1.08 );                         // Price
    row.SetDate(      6, 2029, 1, 1 );                   // Expiration date
    row.SetCharString(7, "Bohnen" );                     // Produkt
    for ( int i = 0; i < 10; ++i ) {
        row.SetInteger(4, i * 10)  ;                     // Quantity
        row.SetBoolean(5, i % 2 == 1 );                  // Taxed
        table->Insert( row );
    }
}

int main( int argc, char* argv[] )
{
    using namespace Tableau;

    try {
        Extract extract( L"order-cpp.tde" );
        std::shared_ptr<Table> table;

        if ( !extract.HasTable( L"Extract" ) ) {
            // Table does not exist; create it.
            TableDefinition tableDef;
            MakeTableDefinition( tableDef );
            table = extract.AddTable( L"Extract", tableDef );
        }
        else {
            // Open an existing table to add more rows.
            table = extract.OpenTable( L"Extract" );
        }

        std::shared_ptr<Tableau::TableDefinition> tableDef = table->GetTableDefinition();
        PrintTableDefinition( *tableDef );

        InsertData( table );
    }
    catch ( const TableauException& e) {
        std::wcerr << L"Something bad happened: " << e.GetMessage() << std::endl;
        exit( EXIT_FAILURE );
    }

    return 0;
}

File: dataprep-transformation/src/main/resources/TDE_API/share/doc/dataextract-8300.15.0308.1149/samples/com/tableausoftware/demos/MakeOrder.java
// -----------------------------------------------------------------------
// The information in this file is the property of Tableau Software and
// is confidential.
//
// Copyright (C) 2012  Tableau Software.
// Patents Pending.
// -----------------------------------------------------------------------
// externalapi/samples/com/tableausoftware/demos/MakeOrder.java
// -----------------------------------------------------------------------
package com.tableausoftware.demos;

import com.tableausoftware.TableauException;
import com.tableausoftware.DataExtract.*;

public final class MakeOrder {

    // Define the table's schema
    private static TableDefinition makeTableDefinition() throws TableauException {
        TableDefinition tableDef = new TableDefinition();
        tableDef.setDefaultCollation(Collation.EN_GB);

        tableDef.addColumn("Purchased",       Type.DATETIME);
        tableDef.addColumn("Product",         Type.CHAR_STRING);
        tableDef.addColumn("uProduct",        Type.UNICODE_STRING);
        tableDef.addColumn("Price",           Type.DOUBLE);
        tableDef.addColumn("Quantity",        Type.INTEGER);
        tableDef.addColumn("Taxed",           Type.BOOLEAN);
        tableDef.addColumn("Expiration Date", Type.DATE);

        // Column with non-default collation
        tableDef.addColumnWithCollation("Produkt", Type.CHAR_STRING, Collation.DE);

        return tableDef;
    }

    // Print a Table's schema to stderr.
    private static void printTableDefinition(TableDefinition tableDef) throws TableauException {
        int numColumns = tableDef.getColumnCount();
        for ( int i = 0; i < numColumns; ++i ) {
            Type type = tableDef.getColumnType(i);
            String name = tableDef.getColumnName(i);

            System.err.format("Column %d: %s (%#06x)\n", i, name, type.getValue());
        }
    }

    // Insert a few rows of data
    private static void insertData(Table table) throws TableauException {
        TableDefinition tableDef = table.getTableDefinition();
        Row row = new Row(tableDef);

        row.setDateTime(  0, 2012, 7, 3, 11, 40, 12, 4550); // Purchased
        row.setCharString(1, "Beans");                      // Product
        row.setString(    2, "uniBeans");                   // uProduct
        row.setDouble(    3, 1.08);                         // Price
        row.setDate(      6, 2029, 1, 1);                   // Expiration date
        row.setCharString(7, "Bohnen");                     // Produkt

        for ( int i = 0; i < 10; ++i ) {
            row.setInteger(4, i * 10);                      // Quantity
            row.setBoolean(5, i % 2 == 1);                  // Taxed
            table.insert(row);
        }
    }

    public static void main( String[] args ) {

        try (Extract extract = new Extract("order-java.tde")) {

            Table table;
            if (!extract.hasTable("Extract")) {
                // Table does not exist; create it
                TableDefinition tableDef = makeTableDefinition();        
                table = extract.addTable("Extract", tableDef);
            }
            else {
                // Open an existing table to add more rows
                table = extract.openTable("Extract");
            }

            TableDefinition tableDef = table.getTableDefinition();
            printTableDefinition(tableDef);

            insertData(table);
        }
        catch (Throwable t) {
            t.printStackTrace(System.err);
        }
    }
}
